### PR TITLE
feat: Enhance GitHub code review with improved context handling and validation

### PR DESCRIPTION
**Pull request by finxo**
# Pull Request

## 📝 Summary
This PR significantly enhances the GitHub code review functionality by improving context handling, adding safety validations, and refining the review workflow. Key improvements include better comment filtering, worktree reference support, enhanced PR classification, and improved error handling. The changes also introduce inline comment validation and rejection classification, along with better handling of pending review conflicts.

## 🔧 Changes Made
- Enhanced GitHub PR thread resolution workflow by removing disabled extension
- Added inline comment safety validation and rejection classification for GitHub reviews
- Implemented worktree reference support and improved context resolution strategy
- Added semantic correctness and state consistency review axes with updated checklist items and scoring logic
- Improved GitHub API error handling with stdout/stderr preservation and enhanced logging
- Enhanced comment filtering and added invalid comment rejection handling
- Added role-based complexity scoring and active review detection for PR classification
- Extracted comment context serialization to shared formatting operations
- Removed unused skill and documentation loading functions from code review operations
- Simplified code review operations by removing build_review_payload and unused operations

## 🧪 Testing
- [x] Unit tests added/updated (`poetry run pytest`)
- [x] All tests passing (`make test`)
- [ ] Manual testing with `titan-dev`

New unit tests were added to cover:
- Findings prompt construction with worktree references
- Comment context building with automated comment filtering
- Change manifest building with file type detection
- Various edge cases in review operations

## 📊 Logs
- [ ] No new log events

## ✅ Checklist
- [x] Self-review done
- [x] Follows the project's [logging rules](.claude/docs/logging.md) (no secrets, no content in logs)
- [x] New and existing tests pass
- [ ] Documentation updated if needed
- [ ] Plugin documentation updated when plugin functions or parameters changed (`Plugins > Git Plugin`, `GitHub Plugin`, `Jira Plugin`)

### CODE REVIEW

File: plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py [hunk 1]
```diff
@@ -1428,9 +2017,71 @@ def submit_review_actions(ctx: WorkflowContext) -> WorkflowResult:
...
+                    ctx.textual.error_text(f"Failed to create review: {err}")
+                    ctx.textual.end_step("error")
+                    return Error(f"Failed to create draft review: {err}")
+            else:
+                ctx.textual.error_text(f"Failed to create review: {err}")
+                ctx.textual.end_step("error")
+                return Error(f"Failed to create draft review: {err}")
 
     with ctx.textual.loading("Submitting review..."):
         submit_result = ctx.github.submit_review(pr_number, review_id, event, review_body)
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py]: The `body` parameter in the final `submit_review` call should probably be `payload.get("body", "")` instead of `review_body`. 

Currently, if `build_review_action_payload` contributes a summary to the review body, it gets overwritten on submission because this call only passes the manually entered `review_body`. This would cause the AI-generated summary to be lost.

File: plugins/titan-plugin-github/titan_plugin_github/models/validators.py [hunk 2]
```diff
@@ -49,124 +27,84 @@ def __init__(
...
-    sim = SequenceMatcher(
+    similarity = SequenceMatcher(
         None,
         new_finding.title.lower(),
         existing.title.lower(),
     ).ratio()
-    return sim > title_similarity_threshold
 
+    same_category = new_finding.category.lower() == (existing.category or "").lower()
+    if same_category and not existing.is_resolved:
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/models/validators.py:95]: The duplicate detection logic seems to have an issue with how it handles resolved comments. The condition `if same_category and not existing.is_resolved:` will cause the check to fail for findings that have the same category as an already resolved comment.

This could lead to re-reporting issues that have already been fixed and marked as resolved, creating unnecessary noise. Shouldn't we consider a finding a duplicate if it shares the same category, regardless of its resolution status, to avoid this?

File: plugins/titan-plugin-github/tests/operations/test_manifest_operations.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        thread_id="general_3",
+        main_comment=bot_comment,
+        replies=[],
+        is_resolved=False,
+        is_outdated=False,
+    )
+
+    context = build_comment_review_context([], [thread], max_entries=5, max_chars=1000)
+
+    assert context == []
```
Comment (finxo) [inline plugins/titan-plugin-github/tests/operations/test_manifest_operations.py:99]: This test passes because the `author_login` ends with `[bot]`, but the comment body suggests it might also be testing for HTML content filtering. The logic in `_is_automated_comment` only checks the author name. 

To make the tests clearer and less brittle, I'd suggest simplifying this test to focus only on the author name condition, as the HTML content is irrelevant for this specific filtering rule. Another test already covers HTML-based filtering (`test_build_comment_review_context_filters_wiz_html_comments`).

File: plugins/titan-plugin-github/tests/operations/test_manifest_operations.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
+from titan_plugin_github.models.review_enums import CommentContextKind, FileChangeStatus
+from titan_plugin_github.models.view import UIComment, UICommentThread, UIFileChange
+from titan_plugin_github.models.review_models import ExistingCommentIndexEntry, Finding
+from titan_plugin_github.models.review_enums import FindingSeverity
+from titan_plugin_github.models.validators import is_duplicate
```
Comment (finxo) [inline plugins/titan-plugin-github/tests/operations/test_manifest_operations.py:5]: The test `test_is_duplicate_matches_adjudicated_resolved_thread_more_aggressively` covers the happy path for the `is_duplicate` function. However, there are no tests for the negative paths. 

Please add a few more tests to cover cases where `is_duplicate` should return `False`, for example:
- The line numbers are different.
- The file paths are different.
- The existing comment is not `is_adjudicated`.

**Review by finxo — COMMENTED** [review]
**plugins/titan-plugin-github/titan_plugin_github/operations/review_strategy_operations.py**:
This new file contains a lot of critical, deterministic logic for classifying and planning the review. It seems highly testable. Could we add unit tests for `classify_pr`, `score_review_candidates`, and `select_review_strategy` to ensure this core logic is robust and to prevent future regressions?

---

**plugins/titan-plugin-github/titan_plugin_github/models/validators.py**:
I noticed that the validation check for `read_mode: full_file` on large files (`_check_full_file_mode`) has been removed from the `ReviewPlanValidator`. This seems like a regression, as it was a useful safeguard to prevent the AI from requesting to load very large files, which could lead to performance problems or exceed model context limits.

Was the removal intentional? If so, what is the new strategy to handle such cases? If not, I'd recommend we restore this validation.

File: plugins/titan-plugin-github/titan_plugin_github/models/validators.py (hunk 2 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/models/validators.py:95]: This is intentional. Resolved threads are not treated as automatic duplicates unless they were adjudicated; otherwise we still allow resurfacing similar findings when a prior resolution may have been incomplete or incorrect.

File: plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py (hunk 1 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py]: 0f485025

File: plugins/titan-plugin-github/tests/operations/test_manifest_operations.py (hunk 3 above)
Comment (finxo) [reply plugins/titan-plugin-github/tests/operations/test_manifest_operations.py:99]: 4c6468d2

File: plugins/titan-plugin-github/tests/operations/test_manifest_operations.py (hunk 4 above)
Comment (finxo) [reply plugins/titan-plugin-github/tests/operations/test_manifest_operations.py:5]: 188ba71e

File: plugins/titan-plugin-github/titan_plugin_github/models/validators.py [hunk 5]
```diff
@@ -49,124 +27,84 @@ def __init__(
         self.offered_checklist_ids = offered_checklist_ids or _ALL_CHECKLIST_IDS
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/models/validators.py]: The use of `or` for defaulting `offered_checklist_ids` can lead to incorrect validation behavior. If an empty `frozenset` is passed to the constructor, it will be treated as `False`, causing `self.offered_checklist_ids` to be populated with `_ALL_CHECKLIST_IDS` instead of remaining empty.

This would allow the AI to propose review axes even when none were offered. To fix this, you should use an explicit `is None` check to handle the default case, ensuring that an intentionally empty set of offered IDs is respected.

```python
self.offered_checklist_ids = (
    _ALL_CHECKLIST_IDS if offered_checklist_ids is None else offered_checklist_ids
)
```

File: plugins/titan-plugin-github/titan_plugin_github/models/validators.py [hunk 6]
```diff
@@ -49,124 +27,84 @@ def __init__(
...
+    if same_category and not existing.is_resolved:
+        return True
 
-def _lines_are_close(
-    line_a: int | None,
-    line_b: int | None,
-    window: int,
-) -> bool:
-    """Return True if both lines are within proximity window, or both are file-level."""
+    if same_category and existing.is_adjudicated:
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/models/validators.py]: This condition for detecting duplicates seems overly aggressive and could lead to the suppression of valid new findings. It marks a finding as a duplicate if it's near an adjudicated comment of the same category, regardless of whether the findings' topics are actually related.

This could cause a valid new issue (e.g., a high-complexity problem) to be suppressed simply because a different, minor issue of the same category (e.g., 'performance') was previously dismissed nearby.

Consider incorporating title similarity into this check as well to ensure you're only suppressing new findings that are genuinely re-raising an already-adjudicated issue.

File: plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py [hunk 7]
```diff
@@ -94,6 +104,251 @@ def _log_ai_response(step_name: str, cli_name: str, stdout: str, stderr: str, ex
...
+        update={
+            "prompt_actual_chars": actual_chars,
+            "prompt_still_too_large": True,
+            "degraded_context": True,
+        }
+    )
+    return [oversized], False
+
+
+def _filter_invalid_inline_comments(ctx: WorkflowContext, pr_number: int, payload: dict) -> tuple[dict, list[dict]]:
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py:278]: This function `_filter_invalid_inline_comments` is a creative way to work around GitHub's review submission limitations, but it appears to be unused in the current workflow. If it were active, its behavior of making two API calls per comment would likely lead to performance issues and hitting API rate limits on reviews with many comments. Since it's currently dead code, it should probably be removed to avoid confusion.

File: plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py [hunk 8]
```diff
@@ -101,11 +101,53 @@ def get_hunks(self, path: str) -> list[ParsedHunk]:
...
+        if not hunks:
+            return []
+
+        file_lines = file_content.split("\n")
+        expanded: list[str] = []
+
+        for hunk in hunks:
+            expand_start = max(0, hunk.new_line_start - extra_lines - 1)
+            expand_end = min(len(file_lines), hunk.new_line_end + extra_lines)
+            surrounding = "\n".join(file_lines[expand_start:expand_end])
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py:130]: The `build_expanded_hunks` function generates a `surrounding` context that includes all lines within the hunk's range, plus some extra lines. However, the original hunk content is appended afterwards, which means that context lines (lines unchanged in the diff) are included twice: once in the `surrounding` block (without a ` ` prefix) and again in the `diff hunk` block (with a ` ` prefix). This duplication makes the resulting context larger than necessary. The surrounding context should ideally only contain lines *outside* the original hunk.

File: plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py [hunk 9]
```diff
@@ -569,4 +640,39 @@ def build_focused_diff_from_hunk(
...
 
 
+def _extract_best_anchor_from_text(text: Optional[str]) -> Optional[str]:
+    """Extract a short single-line anchor from evidence text."""
+    if not text:
+        return None
+
+    for raw_line in text.splitlines():
+        line = raw_line.strip()
+        if not line or line.startswith("```") or line.startswith("//"):
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py]: The logic to identify and skip comment lines in `_extract_best_anchor_from_text` only handles `//` comments. This makes it language-specific and it will fail for languages like Python, Ruby, or shell scripts that use `#`. To make this more robust, please consider handling a wider range of common comment prefixes.

File: plugins/titan-plugin-github/titan_plugin_github/operations/review_action_operations.py [hunk 10]
```diff
@@ -88,46 +105,93 @@ def build_review_action_payload(
             continue
 
         # new_comment — try inline first, fall back to general body
-        if action.path and action.line:
+        resolved_line = action.resolved_line
+        if manager and action.path:
+            resolved_line = resolved_line or manager.resolve_line_anchor(
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/operations/review_action_operations.py]: The line anchor resolution logic is being re-executed within `build_review_action_payload`. This is problematic because this function should operate on actions that have already been fully resolved by `resolve_action_anchors`.

This re-resolution can lead to an inconsistent state. For example, if a line anchor is resolved here, an inline comment might be posted, but the `action` object itself is not updated. Consequently, its metadata fields like `is_inline_safe_for_github` and `resolved_line` become stale, and the detailed debugging logs at the end of the function will be misleading.

To ensure data consistency and make the control flow clearer, `build_review_action_payload` should rely solely on the pre-existing `action.resolved_line` value. If it's `None`, the existing fallback to a general review comment is the correct and safe behavior.

Please remove the redundant call to `manager.resolve_line_anchor(...)`. A similar issue is present in `extract_diff_hunk_for_action`.

```python
        # This re-resolution should be removed.
        if manager and action.path:
            resolved_line = resolved_line or manager.resolve_line_anchor(
                action.path,
                line=action.line,
                snippet=action.anchor_snippet,
                evidence=action.evidence,
            )
```
